# Regressed Classification Weekend Project

### **Challenge Options**

#### Saudi Airbnb Price Prediction
- Data: [github.com/gumdropsteve/intro_to_machine_learning/tree/main/day_10/data](https://github.com/gumdropsteve/intro_to_machine_learning/tree/main/day_10/data)
- Data Dictionary: [github.com/gumdropsteve/intro_to_machine_learning/tree/main/day_10/data/README.MD](https://github.com/gumdropsteve/intro_to_machine_learning/tree/main/day_10/data/README.MD)
- Type: regression
- Starter code: [gist.github.com/gumdropsteve/c178891652838cb0663a4ef1d2eb9d14](https://gist.github.com/gumdropsteve/c178891652838cb0663a4ef1d2eb9d14)
- Target: `price`

#### Income Bracket Prediction
- Data: [archive.ics.uci.edu/ml/machine-learning-databases/adult/](https://archive.ics.uci.edu/ml/machine-learning-databases/adult/)
- Data Dictionary: [archive.ics.uci.edu/ml/datasets/Adult](https://archive.ics.uci.edu/ml/datasets/Adult)
- Type: classification
- Starter code: [gist.github.com/gumdropsteve/25e40b0ab0b5a6a7d51f11c00f91d0bb](https://gist.github.com/gumdropsteve/25e40b0ab0b5a6a7d51f11c00f91d0bb)
- Target: `target`

#### Mushroom
- Data: [archive.ics.uci.edu/ml/machine-learning-databases/mushroom/](https://archive.ics.uci.edu/ml/machine-learning-databases/mushroom/)
- Data Dictionary: [archive.ics.uci.edu/ml/datasets/mushroom](https://archive.ics.uci.edu/ml/datasets/mushroom)
- Type: classification
- Starter code: [colab.research.google.com/drive/1hIoS2jZCpVOCE4U2mxqkeYPLh2m6I8rA?usp=sharing](https://colab.research.google.com/drive/1hIoS2jZCpVOCE4U2mxqkeYPLh2m6I8rA?usp=sharing)
- Target: `target`

### **Available Models**
Pick from these models based on which challenge you take on (regression or classification). 

#### Regression
- [LinearRegression](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html)
- [SGDRegressor](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.SGDRegressor.html)
- [KNeighborsRegressor](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsRegressor.html)

#### Classification
- [LogisticRegression](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html)
- [SGDClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.SGDClassifier.html)
- [KNeighborsClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html)

## **Deleverables**
1. EDA
   - What is this data?
   - Visualize some relationships
   - What do you see? What did you find?
2. Data prep & `train_test_split()`
    - Make this data ready for a ML model
    - Decide which features to include
    - Split into training and testing datasets
3. Baseline model
    - What is your baseline model?
        - Explain what's going on please... Assume I'm an idiot 
4. Create a model
    - Compare your results to the baseline
5. Optimize your model
    - [GridSearchCV](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html)
        - Bonus: [RandomizedSearchCV](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.RandomizedSearchCV.html)
    - Compare your results to the baseline
        - How did you do this? Why?
    - If you don't beat the baseline, try improving your model or using a different model
        - Were you able to beat the baseline? Why?

In [ ]:
import pandas as pd

dammam_link = 'https://github.com/gumdropsteve/intro_to_machine_learning/raw/main/day_10/data/dammam.parquet'
jeddah_link = 'https://github.com/gumdropsteve/intro_to_machine_learning/raw/main/day_10/data/jeddah.parquet'
riyadh_link = 'https://github.com/gumdropsteve/intro_to_machine_learning/raw/main/day_10/data/riyadh.parquet'

df = pd.read_parquet(riyadh_link)

df

ds  ... chefs_kitchen_bool
0      2021-01-20 04:14:08.745592  ...              False
1      2021-01-20 04:14:08.745592  ...              False
2      2021-01-20 04:14:08.745592  ...              False
3      2021-01-20 04:14:08.745592  ...              False
4      2021-01-20 04:14:08.745592  ...              False
...                           ...  ...                ...
11765  2021-02-01 06:19:53.258868  ...              False
11766  2021-02-01 06:19:53.258868  ...              False
11767  2021-02-01 06:19:53.258868  ...              False
11768  2021-02-01 06:19:53.258868  ...              False
11769  2021-02-01 06:19:53.258868  ...              False

[11770 rows x 40 columns]

## Simple Solve
Getting a MVP out the door...

In [ ]:
corr = df.corr()

corr = corr.head(10)

corr.style.background_gradient(cmap='coolwarm')

/usr/local/lib/python3.6/dist-packages/pandas/io/formats/style.py:1126: RuntimeWarning: All-NaN slice encountered
  smin = np.nanmin(s.to_numpy()) if vmin is None else vmin
/usr/local/lib/python3.6/dist-packages/pandas/io/formats/style.py:1127: RuntimeWarning: All-NaN slice encountered
  smax = np.nanmax(s.to_numpy()) if vmax is None else vmax


In [ ]:
# display values from row 7 (price)
corr.iloc[7]

guests                    -0.001213
bedrooms                   0.015028
beds                      -0.001213
is_studio                  0.040343
baths                      0.340305
half_baths                      NaN
shared_baths               0.301257
price                      1.000000
avg_rating                 0.154220
n_reviews                 -0.006790
gym_bool                   0.236375
wifi_bool                  0.043523
self_check_in_bool        -0.046619
air_conditioning_bool     -0.015725
pets_allowed_bool         -0.040202
indoor_fireplace_bool      0.055843
hot_tub_bool              -0.018637
free_parking_bool         -0.039740
pool_bool                  0.138494
kitchen_bool              -0.050713
breakfast_bool            -0.032542
elevator_bool             -0.055278
washer_bool               -0.032167
dryer_bool                -0.023855
heating_bool                    NaN
waterfront_bool           -0.009394
dishwasher_bool            0.025785
beachfront_bool           -0

In [ ]:
import numpy as np

np.sum(df.isnull())

ds                             0
search_filter               3000
url                            1
title                          0
type                           0
location                       0
guests                      6150
bedrooms                       0
beds                        6150
is_studio                      0
baths                       1828
half_baths                 11723
shared_baths               11514
price                          0
avg_rating                  5701
n_reviews                   5701
gym_bool                       0
wifi_bool                      0
self_check_in_bool             0
air_conditioning_bool          0
pets_allowed_bool              0
indoor_fireplace_bool          0
hot_tub_bool                   0
free_parking_bool              0
pool_bool                      0
kitchen_bool                   0
breakfast_bool                 0
elevator_bool                  0
washer_bool                    0
dryer_bool                     0
heating_bo

In [ ]:
from sklearn.model_selection import train_test_split

X = df[['bedrooms', 'baths', 'shared_baths', 'avg_rating', 'n_reviews', 
        'gym_bool', 'pool_bool', 'dishwasher_bool']]
y = df.price

# all missing values here likely mean None, so let's fill with 0
X = X.fillna(0)

X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.8)

In [ ]:
len(X_train), len(y_test)

(9416, 2354)

In [ ]:
from sklearn.linear_model import LinearRegression

lr = LinearRegression()

lr.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [ ]:
preds = lr.predict(X_test)

# show the last 5 predictions
preds[-5: ]

array([179.87423474, 369.71005068, 147.21650049, 302.27088217,
       225.52129746])

In [ ]:
from sklearn.metrics import r2_score

r2_score(y_test, preds)

0.13390662287969268

#### Adding a quick categorical feature to that...
Turning `type` into numbers because I have a feeling it might help..

In [ ]:
df.type.unique()

array(['Entire chalet', 'Entire apartment', 'Room', 'Entire condominium',
       'Entire guest suite', 'Entire villa', 'Entire house',
       'Private room', 'Tiny house', 'Resort room', 'Entire loft',
       'Entire serviced apartment', 'Hotel room', 'Campsite', 'Camper/RV',
       'Hut', 'Farm stay', 'Dome house', 'Earth house',
       'Entire guesthouse', 'Entire place', 'Entire cottage',
       'Shared room', 'Tent'], dtype=object)

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.get_dummies.html

In [ ]:
pd.get_dummies(df.type)

Camper/RV  Campsite  Dome house  ...  Shared room  Tent  Tiny house
0              0         0           0  ...            0     0           0
1              0         0           0  ...            0     0           0
2              0         0           0  ...            0     0           0
3              0         0           0  ...            0     0           0
4              0         0           0  ...            0     0           0
...          ...       ...         ...  ...          ...   ...         ...
11765          0         0           0  ...            0     0           0
11766          0         0           0  ...            0     0           0
11767          0         0           0  ...            0     0           0
11768          0         0           0  ...            0     0           0
11769          0         0           0  ...            0     0           0

[11770 rows x 24 columns]

In [ ]:
X = df[['bedrooms', 'baths', 'shared_baths', 'avg_rating', 'n_reviews', 'gym_bool', 'pool_bool', 'dishwasher_bool']]

pd.concat([X, pd.get_dummies(df['type'])], axis=1)

bedrooms  baths  shared_baths  ...  Shared room  Tent  Tiny house
0           1.0    1.0           NaN  ...            0     0           0
1           1.0    1.0           NaN  ...            0     0           0
2           1.0    1.0           NaN  ...            0     0           0
3           1.0    1.0           NaN  ...            0     0           0
4           1.0    1.5           NaN  ...            0     0           0
...         ...    ...           ...  ...          ...   ...         ...
11765       1.0    2.0           NaN  ...            0     0           0
11766       2.0    3.0           NaN  ...            0     0           0
11767       2.0    2.0           NaN  ...            0     0           0
11768       2.0    2.5           NaN  ...            0     0           0
11769       1.0    1.5           NaN  ...            0     0           0

[11770 rows x 32 columns]

In [ ]:
X = pd.concat([X, pd.get_dummies(df['type'])], axis=1)
y = df.price

# fill missing values in X with 0
X = X.fillna(0)

X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.8)

In [ ]:
from sklearn.linear_model import LinearRegression

lr = LinearRegression()

lr.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [ ]:
preds = lr.predict(X_test)

r2_score(y_test, preds)

0.2010273679837733

It helped a little! :)

## Data Wrangling
Improving the quality of the data...

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10120 entries, 0 to 10119
Data columns (total 40 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ds                       10120 non-null  object 
 1   search_filter            6820 non-null   object 
 2   url                      10120 non-null  object 
 3   title                    10120 non-null  object 
 4   type                     10120 non-null  object 
 5   location                 10120 non-null  object 
 6   guests                   6118 non-null   float64
 7   bedrooms                 10120 non-null  float64
 8   beds                     6118 non-null   float64
 9   is_studio                10120 non-null  bool   
 10  baths                    8585 non-null   float64
 11  half_baths               64 non-null     float64
 12  shared_baths             209 non-null    float64
 13  price                    10120 non-null  float64
 14  avg_rating            

In [ ]:
df.describe()

guests      bedrooms  ...   avg_rating    n_reviews
count  6118.000000  10120.000000  ...  4927.000000  4927.000000
mean      2.603465      1.488834  ...     4.719351    19.680536
std       1.788836      1.043132  ...     0.326824    23.585304
min       0.000000      0.000000  ...     3.330000     3.000000
25%       2.000000      1.000000  ...     4.640000     5.000000
50%       2.000000      1.000000  ...     4.790000    11.000000
75%       3.000000      2.000000  ...     4.960000    25.000000
max      19.000000     10.000000  ...     5.000000   154.000000

[8 rows x 9 columns]

`url` can be used to identify a listing. As no 2 listings could share the same URL, each unique value in `url` represents 1 unique listing.. 

Source: [Data dictionary](https://github.com/gumdropsteve/intro_to_machine_learning/blob/main/day_10/data/README.md#data-dictionary)

In [ ]:
len(df['url'].unique())

342

Because we have 342 unique values in `url`, we have 342 unique listings...

To avoid data leakage, we need to only keep 1 instance of each listing (`url`)... So let's drop the `ds` (datetime the info was pulled) and `search_filter` (search filter applied on top of the city filter) because the same listing (`url`) could show up multiple times under different `ds` and `search_filter` values...

In [ ]:
len(df.drop(['ds', 'search_filter'], axis=1).drop_duplicates())

1874

That didn't cut it down to 342... But it's a start.

In [ ]:
dropped_df = df.drop(['ds', 'search_filter'], axis=1).drop_duplicates()

dropped_df.sample(3)

url  ... chefs_kitchen_bool
7384  http://airbnb.com/rooms/42599774  ...              False
649   http://airbnb.com/rooms/42293627  ...              False
627   http://airbnb.com/rooms/43443981  ...              False

[3 rows x 38 columns]

In [ ]:
len(dropped_df['url'].unique())

342

Let's add back the `ds` and `search_filter` columns and check out some of the `url` values that are still seen more than once..

In [ ]:
df[['ds', 'search_filter']].iloc[dropped_df.index]

ds  search_filter
0      2021-01-20 04:13:49.541603           None
1      2021-01-20 04:13:49.541603           None
2      2021-01-20 04:13:49.541603           None
3      2021-01-20 04:13:49.541603           None
4      2021-01-20 04:13:49.541603           None
...                           ...            ...
9937   2021-02-01 05:23:08.769406    hotel_rooms
9988   2021-02-01 05:23:23.169889  private_rooms
10014  2021-02-01 05:23:25.692193  private_rooms
10031  2021-02-01 05:23:29.380056  private_rooms
10046  2021-02-01 05:23:40.290436   shared_rooms

[1874 rows x 2 columns]

In [ ]:
dropped_df[['ds', 'search_filter']] = df[['ds', 'search_filter']].iloc[dropped_df.index]

dropped_df.sample(3)

url  ... search_filter
6820  http://airbnb.com/rooms/46428784  ...  entire_homes
4747  http://airbnb.com/rooms/36351268  ...          None
7060  http://airbnb.com/rooms/43140633  ...  entire_homes

[3 rows x 40 columns]

In [ ]:
non_unique_urls = []

# go through the unique values in the url column
for _ in dropped_df.url.unique():
    # if this url is seen more than once
    if len(dropped_df.loc[dropped_df['url'] == _]) > 1:
        # add it to the non-unique urls list
        non_unique_urls.append(_)

len(non_unique_urls)

331

In [ ]:
print(f'{len(non_unique_urls) / len(df.url.unique()) * 100}% of values in `url` are still not unique\nthis means those rows have the same value in `url` but different values in at least 1 column that is not `ds`, `search_filter` or `url`')

96.78362573099415% of values in `url` are still not unique
this means those rows have the same value in `url` but different values in at least 1 column that is not `ds`, `search_filter` or `url`


In [ ]:
# what is the 0th non-unique URL
non_unique_urls[0]

'http://airbnb.com/rooms/47145509'

In [ ]:
# how many times is the 0th non-unique URL seen?
len(dropped_df.loc[dropped_df['url'] == non_unique_urls[0]])

12